app/services/course/assessment/answer/text_response_auto_grading_service.rb

```diff
@@ -0,0 +1,73 @@
+class Course::Assessment::Answer::TextResponseAutoGradingService < \
+  Course::Assessment::Answer::AutoGradingService
+  def grade(auto_grading)
+    auto_grading.answer.grade, messages = grade_answer(auto_grading.answer.actable)
+    auto_grading.result = { messages: messages }
+    super(auto_grading)
+  end
+
+  private
+
+  # Grades the given answer.
+  #
+  # @param [Course::Assessment::Answer::TextResponse] answer The answer specified by the
+  #   student.
+  # @return [Array<(Fixnum, Object)>] The grade and the messages to be assigned to the grading.
+  def grade_answer(answer)
+    question = answer.question.actable
+    answer_text = answer.answer_text
+    exact_matches, keywords = question.solutions.partition(&:exact_match?)
+
+    solutions = find_exact_match(answer_text, exact_matches)
+    # If there is no exact match, we fall back to keyword matches.
+    # Solutions are always kept in an array for easier use of #grade_for and #explanations_for
+    if !solutions.nil?
+      solutions = [solutions]
+    else
+      solutions = find_keywords(answer_text, keywords)
+    end
+
+    [grade_for(question, solutions), explanations_for(solutions)]
+  end
+
+  # Returns one solution that exactly matches the answer.
+  #
+  # @param [String] answer_text The answer text entered by the student.
+  # @param [Array<Course::Assessment::Question::TextResponseSolution>] solutions The solutions
```

Comment:
this is @param, not the @return. is @return needed? since it's described by the one line summary.